### ISSUE TITLE
Add document to support Topological based Provisioning In Internal Mode

### ISSUE THREAD

**Issue by parth-gr** [rook/rook#14182]
**Is this a bug report or feature request?**
* Feature Request

Add document to support Topological based Provisioning In Internal Mode,
Similar we have for external mode
 https://rook.github.io/docs/rook/v1.14/CRDs/Cluster/external-cluster/topology-for-external-mode/

**What should the feature do:**

DO topological based provisioning

**What is use case behind this feature:**

Help in utilizing replica-1 pools

**Environment**:

**Comment by Madhu-1** [comment]
Reopening as its good to have documentation for it.

**Comment by travisn** [comment]
@fyang13 The steps to configure this are already possible today, as follows:
1. For each zone
   - Create OSDs in the zone that specify their `deviceClass` property, so that only the replica 1 pools will store data on these OSDs
   - Create a CephBlockPool with replica 1 and with a `deviceClass` that matches the OSDs
2. Create a storage class with the `topologyConstraintPools` json similar to [this example](https://rook.github.io/docs/rook/v1.14/CRDs/Cluster/external-cluster/topology-for-external-mode/#kubernetes-cluster). The CSI driver will now provision PVs in the same zone as the application, with the corresponding replica 1 pool. 
3. Set the `deviceClass` on all other pools in the cluster (if deviceClass is set on any pool, it must be set on all pools or else ceph raises health warnings). If you have existing OSDs that are already used in other replicated pools, their device class is likely `ssd`. The device class of the OSDs can be seen with commands such as `ceph osd tree`.

**Comment by fyang13** [comment]
@travisn Yes, I figure it would work, just try to get the helm-chart working, I assume I can add multiple CephBlockPool here, but how do I associate it to different zones? I see the storage definition [here](https://github.com/rook/rook/blob/597e5e9be7d731cb140f952759b34b3c2b9845cd/deploy/charts/rook-ceph-cluster/values.yaml#L358) is at cluster level, how do I define different deviceClass here for OSDs in different zones?

**Comment by travisn** [comment]
If you are deploying with the [cluster-on-pvc](https://rook.io/docs/rook/latest-release/CRDs/Cluster/pvc-cluster/), you can define the storageClassDeviceSets with affinity to a zone, and set the [crushDeviceClass](https://github.com/rook/rook/blob/master/deploy/examples/cluster-on-pvc.yaml#L127-L131). 

If you are deploying with the [host-based cluster](https://rook.io/docs/rook/latest-release/CRDs/Cluster/host-cluster/), you would need to define the `nodes` and then set `config.deviceClass` under a specific device.

**Comment by fyang13** [comment]
We are deploying it with host-based cluster. Instead of defining nodes and deviceClass, would it work by just adding multiple `storageClassDeviceSets` under `cephClusterSpec.storage` section? e.g.:
`storage:
    onlyApplyOSDPlacement: false
    allowOsdCrushWeightUpdate: false
    useAllNodes: true
    useAllDevices: false
    deviceFilter: "^nvme"
    storageClassDeviceSets:
      - name: set1
        placement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone1
        preparePlacement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone1
      - name: set2
        placement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone2
        preparePlacement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone2
      - name: set3
        placement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone3
        preparePlacement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: topology.kubernetes.io/zone
                      operator: In
                      values:
                        - zone3
`

**Comment by travisn** [comment]
If you have different devices classes on each node, it will certainly be tedious to use the `nodes` schema. Using `storageClassDeviceSets` does require PVs. See the example [cluster-on-local-pvc.yaml](https://github.com/rook/rook/blob/master/deploy/examples/cluster-on-local-pvc.yaml) to see how to define the devices as local PVs and then you can use the storageClassDeviceSets.

**Comment by fyang13** [comment]
I'm having difficulty to get this working, under the storage section of cephCluster, we have:

`  storage:
    deviceFilter: ^nvme
    flappingRestartIntervalHours: 0
    nodes:
..
    - name: tmirs-rr-957
       config:
        deviceClass: nvme_rr
    store: {}
    useAllDevices: false
    useAllNodes: false
  waitTimeoutForHealthyOSDInMinutes: 10
  `
  
  And `rookcmd` in osd preparer seems pickup the deviceClass override in `osd-crush-device-class` field:
  `2024-11-19 18:52:20.681179 I | rookcmd: flag values: --cluster-id=aa8b6d4d-46e7-4b3c-9385-20cb23d56826, --cluster-name=mir, --data-device-filter=^nvme, --data-device-path-filter=, --data-devices=, --encrypted-device=false, --force-format=false, --help=false, --location=, --log-level=DEBUG, --metadata-device=, --node-name=tmirs-rr-957, --osd-crush-device-class=nvme_rr, --osd-crush-initial-weight=, --osd-database-size=0, --osd-store-type=bluestore, --osd-wal-size=576, --osds-per-device=1, --pvc-backed-osd=false, --replace-osd=-1`
  
 But in the end, `cephosd` is still using the OS device type:
 `
 2024-11-19 18:52:23.051865 I | cephosd: setting device class "nvme" for device "/dev/nvme0n1"
 `
 
This looks very similar as [this issue](https://github.com/rook/rook/issues/11826), but we are running Rook version v1.13.8 and it should have the fix already?

**Comment by travisn** [comment]
Yes, that fix would be in v1.10 and newer, so that should not be the issue.

Did the OSDs already exist and you are trying to update the device class? IIRC, existing OSDs will not update their device class. Or even with new OSDs you are seeing that the device class is not applied? 

@satoru-takeuchi Could you take a look?

**Comment by satoru-takeuchi** [comment]
@travisn Sure.

**Comment by fyang13** [comment]
I wiped the disk, /var/lib/rook dir, and then `ceph auth del osd.{num}`, then restart ceph-operator, voila! The device class show up correctly now. @travisn Thanks for your help!
